csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -13,117 +13,291 @@
 #include <ir/iostream.h>
 #include <ir/utils.h>
 #include <multidevice/utils.h>
+#include <scheduler/utils.h>
+#include <transform_replay.h>
 
 namespace nvfuser::preseg_passes {
 
 namespace {
-void validateMeshes(Fusion* fusion) {
-  // Validate that meshes are assigned to all TensorViews or none.
-  TensorView* tv_with_mesh = nullptr;
-  TensorView* tv_without_mesh = nullptr;
-  for (TensorView* tv : fusion->allTvs()) {
-    auto update_if_null = [](TensorView*& lhs, TensorView* rhs) {
-      if (lhs == nullptr) {
-        lhs = rhs;
-      }
-    };
 
-    if (tv->isCpuScalar()) {
-      continue;
+template <typename Range>
+std::vector<TensorView*> filterTvsWithMesh(const Range& tvs) {
```

Comment:
Time to consider std::ranges::views::filter

Again, can be in a separate PR

Reply:
Will do in a follow up: https://github.com/NVIDIA/Fuser/issues/4298.